Userland/Libraries/LibAudio/PlaybackStreamSerenity.cpp

```diff
@@ -0,0 +1,115 @@
+    : m_connection(move(stream))
+    , m_polling_timer(move(polling_timer))
+    , m_data_request_callback(move(data_request_callback))
+{
+    // Ensure that our audio buffers are filled when they are more than 3/4 empty.
+    // FIXME: Add an event to ConnectionToServer track the sample rate and update this interval.
+    //        Or use a shared condition variable to wake an enqueuing thread to provide data to the audio server.
+    m_polling_timer->set_interval(static_cast<int>((AUDIO_BUFFERS_COUNT * 3 / 4) * AUDIO_BUFFER_SIZE * 1000 / m_connection->get_self_sample_rate()));
+    m_polling_timer->on_timeout = [this]() {
+        fill_buffers();
+    };
+    m_polling_timer->start();
+}
+
+void PlaybackStreamSerenity::fill_buffers()
+{
+    // The number below is picked arbitrarily.
+    // FIXME: This shouldn't need to poll. Ideally, we should be able to recieve a signal when the server
+    //        consumes a buffer and enqueue some data automatically.
+    while (m_connection->can_enqueue()) {
+        Array<Sample, AUDIO_BUFFER_SIZE> buffer;
+        buffer.fill({ 0.0f, 0.0f });
+        auto written_data = m_data_request_callback(Bytes { reinterpret_cast<u8*>(buffer.data()), sizeof(buffer) }, PcmSampleFormat::Float32, AUDIO_BUFFER_SIZE);
+        // FIXME: We should make it so that we can enqueue the exact number of samples we receive from the callback.
+        //        Otherwise, the number of samples that are reported by the server will not be correct for the
+        //        client of this implementation. Eventually, we will want to use server-side timing information, which
+        //        should be accurate down to single samples.
+        m_number_of_samples_enqueued += written_data.size() / sizeof(Sample);
+        MUST(m_connection->realtime_enqueue(buffer));
+    }
+}
+
+void PlaybackStreamSerenity::set_underrun_callback(Function<void()> callback)
+{
+    // FIXME: Implement underrun callback in AudioServer
+    (void)callback;
+}
+
+NonnullRefPtr<Core::ThreadedPromise<Duration>> PlaybackStreamSerenity::resume()
+{
+    auto promise = Core::ThreadedPromise<Duration>::create();
+    // FIXME: We need to get the time played at the correct time from the server. If a message to
+    //        start playback is sent while there is any other message being processed, this may end
+    //        up being inaccurate.
+    auto time = MUST(total_time_played());
+    fill_buffers();
+    m_connection->async_start_playback();
+    m_polling_timer->start();
+    promise->resolve(move(time));
+    return promise;
+}
+
+NonnullRefPtr<Core::ThreadedPromise<void>> PlaybackStreamSerenity::drain_buffer_and_suspend()
+{
+    // FIXME: Play back all samples on the server before pausing. This can be achieved by stopping
+    //        enqueuing samples and receiving a message that a buffer underrun has occurred.
```

Comment:
I think we should achieve that another way: by having two pause modes, one for instant pause, and one for next-opportunity-pause. That can simply be a parameter in the IPC call, and the server is very good at handling this case on its side without ever needing to involve the client.

Reply:
That could work, but the thing is, we also need to know when the pause-on-underrun finishes to resolve the promise. I'm not sure if that's achievable currently with the way the generated IPC functions work?

I don't really consider pumping the event loop a valid solution, as that can result in stack overflows, which is already a bug in LibWeb that has yet to be fixed, due to issues with the existing audio codec plugins.